📖 Add tutorial for external plugin (#3488)
kubernetes-sigs/kubebuilder: docs/book/src/plugins/external-plugins.md

# Extending Kubebuilder with external plugins

## Overview

> You can extend the Kubebuilder functionality by introducing external plugin.

Reviewer: Nit: I would reword this a bit to be more along the lines of:
```suggestion
Kubebuilder's functionality can be extended through the use of external plugins.
```

> An external plugin is an executable that provides the same kubebuilder plugin features. Since it is an executable, such customized plugin can be implemented in any language.

Reviewer 2: Could you elaborate a bit more on what you mean by same kubebuilder plugin features? It might be nice to include those here.

Reviewer: +1 to more elaboration. I would also recommend a bit of a rephrase here to be something more along the lines of:
```suggestion
An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language. 
```

The Kubebuilder CLI loads the external plugin in the specified path and interact the plugin through the executable's `stdin` & `stdout`.

## When is it useful?

> - If you want to create helpers or addons on top of the scaffolds done by the Kubebuilder internal plugins.

Reviewer 2: Should we still use the terminology as "internal plugins" since all plugins will be implemented as external plugins?

Reviewer: +1 on avoiding "internal plugins" as terminology here. I would probably say something more along the lines of "Kubebuilder's default scaffolding"

- If you design customized layouts and want to take advantage of functions from Kubebuilder library.

> - If you are looking for implementing plugins by a different language other than `Go`.

Reviewer: ```suggestion
- If you are looking for implementing plugins in a language other than `Go`.
```

type PluginResponse struct {
	// APIVersion defines the versioned schema of the PluginResponse that is back sent back to Kubebuilder.
	// Initially, this will be marked as alpha (v1alpha1)
	APIVersion string `json:"apiVersion"`

	// Command holds the command that gets executed by the plugin such as init, create api, etc.
	Command string `json:"command"`

	// Metadata contains the plugin specific help text that the plugin returns to Kubebuilder when it receives
	// `--help` flag from Kubebuilder.
	Metadata plugin.SubcommandMetadata `json:"metadata"`

	// Universe in the PluginResponse represents the updated file contents that was written by the plugin.
	Universe map[string]string `json:"universe"`

	// Error is a boolean type that indicates whether there were any errors due to plugin failures.
	Error bool `json:"error,omitempty"`

	// ErrorMsgs contains the specific error messages of the plugin failures.
	ErrorMsgs []string `json:"errorMsgs,omitempty"`

	// Flags contains the plugin specific flags that the plugin returns to Kubebuilder when it receives
	// a request for a list of supported flags from Kubebuilder
	Flags []Flag `json:"flags,omitempty"`
}

> ```

Reviewer: It might be nice to include an example of the JSON for both of these. It might be difficult for someone reading the docs to visualize the JSON based on Go type definitions.

Author: Good idea!
I added sample JSONs for both request & response together with explainations.
Also, the Go definitions are now referred in the link instead of directly pasting them in the docs.

Hope this can make it clearer..

<aside class="note">
<h1>Caution</h1>

When writing your own external plugin, you **should not** directly echo or print anything to the stdout. 

This is because Kubebuilder and your plugin are communicating with each other via `stdin` and `stdout` using structured `JSON` data.
Any additional information sent to stdout (such as debug messages or logs) that's not part of the expected PluginResponse JSON structure may cause parsing errors when Kubebuilder tries to read and decode the response from your plugin.

If you need to include logs or debug messages while developing your plugin, consider writing these messages to a log file instead.

</aside>

## How to use it?

### Prerequisites

- You have built and installed your Kubebuilder binary.

- You have built and installed your external plugin as an **executable / binary**.

> - Your plugin is put in a path following the GVK schema:

Reviewer 2: ```suggestion
- Your plugin is placed in a path following a `group-like name and version` scheme:
```

> - You have built and installed your Kubebuilder binary.
> 
> - You have built and installed your external plugin as an **executable / binary**.
> 
> - Your plugin is put in a path following the GVK schema:
> ```sh
> /path1/path2/${name}/${version} 

Reviewer 2: I like @camilamacedo86's suggestion.
> WDYT about something more like ^ ?

```suggestion
- kubebuilder CLI > 3.11.0
- Your own external plugin executable written in a language of your choice 
- Path of the external plugin configured via the ENV VAR (specify the environment variable here) or it can be placed in a path following a `group-like name and version` scheme:

```

> /path1/path2/${name}/${version} 

Reviewer 2: ```suggestion
 $HOME/.config/kubebuilder/plugins/${name}/${version} for Linux
 ~/Library/Application Support/kubebuilder/plugins/${name}/${version} for OSX
```

```

### Usage:

The external plugin supports the same subcommands as kubebuilder already provides:
- `init`: project initialization.
- `create api`: scaffold Kubernetes API definitions.

> - `create webhook`: scaffold Kubernetes webhooks.

Reviewer 3: it is missing edit

Reviewer: We should also add the optional subcommands for the plugin that provides metadata to Kubebuilder

Also, if the plugin is put in a configured path, you need to use a env variable `$EXTERNAL_PLUGINS_PATH` to tell Kubebuilder where to search for the plugin binary.

Otherwise, Kubebuilder would search for the plugins in a default path based on your OS:
1. On Linux:
```sh
$XDG_CONFIG_HOME/kubebuilder/plugins/${name}/${version}
```

2. On OSX
```sh
~/Library/Application Support/kubebuilder/plugins/${name}/${version}
```

**Examples**
```sh
# Initialize a new project with the external plugin named `sampleplugin`

> EXTERNAL_PLUGINS_PATH=/path1/path2/sampleplugin/v1/sampleplugin kubebuilder init --plugins sampleplugin/v1

Author: Thanks for catching! 
And based on @rashmigottipati 's suggestion, hope using `export EXTERNAL_PLUGINS_PATH=<custom-path>` could be more clear 🤔

# Create a new API with the above external plugin with a customized flag `number`

> EXTERNAL_PLUGINS_PATH=/path1/path2/sampleplugin/v1/sampleplugin kubebuilder create api --plugins sampleplugin/v1 --number 2

Reviewer 2: Instead of `/path1/path2` can we specify something else as it may cause confusion. For example:
```suggestion
EXTERNAL_PLUGINS_PATH=<home-dir>/sampleplugin/v1/sampleplugin kubebuilder create api --plugins sampleplugin/v1 --number 2
```

This comment applies to all other places where `/path1/path2` is mentioned.

# Extending Kubebuilder with external plugins

## Overview

Kubebuilder's functionality can be extended through the use of external plugins.

> An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language. 

Reviewer 4: ```suggestion
An external plugin is an executable (can be implemented in any language) that implements an execution pattern that Kubebuilder knows how to interact with.
```

The Kubebuilder CLI loads the external plugin in the specified path and interacts with it through `stdin` & `stdout`.

## When is it useful?

- If you want to create helpers or addons on top of the scaffolds done by Kubebuilder's default scaffolding.

- If you design customized layouts and want to take advantage of functions from Kubebuilder library.

- If you are looking for implementing plugins in a language other than `Go`.

## How to write it?

The inter-process communication between your external plugin and Kubebuilder is through the standard I/O.

Your external plugin can be written in any language, given it adheres to the [PluginRequest][PluginRequest] and [PluginResponse][PluginResponse] type structures.

`PluginRequest` encompasses all the data Kubebuilder collects from the CLI and previously executed plugins in the plugin chain.
Kubebuilder conveys the marshaled PluginRequest (a `JSON` object) to the external plugin over `stdin`.

> Below is a sample JSON object of the `PluginRequest` type:

Reviewer 4: ```suggestion
Below is a sample JSON object of the `PluginRequest` type, triggered by "kubebuilder init --plugins v1alpha1 --domain my.domain:
```

Author: Thanks for the nice suggestions👍
Just applied!

> Below is a sample JSON object of the `PluginRequest` type, triggered by "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain":

Reviewer: nit:
```suggestion
Below is a sample JSON object of the `PluginRequest` type, triggered by `kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain`:
```

```json
{
    "apiVersion": "v1alpha1",
    "args": ["--domain", "my.domain"],
    "command": "init",
    "universe": {}
}
```

`PluginResponse` represents the updated state of the project, as modified by the plugin. This data structure is serialized into `JSON` and sent back to Kubebuilder via `stdout`.

Here is a sample JSON representation of the `PluginResponse` type:
```json
{
    "apiVersion": "v1alpha1",
    "command": "init",
    "metadata": {
        "description": "The `init` subcommand is meant to initialize a project via Kubebuilder. It scaffolds a single file: `initFile`",

>     	"Examples": "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain"

Reviewer: ```suggestion
    	 "examples": "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain"
```

```sh
# for Linux
$HOME/.config/kubebuilder/plugins/${name}/${version}

# for OSX
~/Library/Application Support/kubebuilder/plugins/${name}/${version}
```

### Subcommands:

The external plugin supports the same subcommands as kubebuilder already provides:
- `init`: project initialization.
- `create api`: scaffold Kubernetes API definitions.
- `create webhook`: scaffold Kubernetes webhooks.
- `edit`: update the project configuration.

Also, there are **Optional** subcommands for a better user experience:
- `metadata`: add customized plugin description and examples when a `--help` flag is specified. 
- `flags`: specify valid flags for Kubebuilder to pass to the external plugin.

<aside class="note">
<h1>More about `flags` subcommand</h1>

The `flags` subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
If a plugin does not implement the `flags` subcommand, Kubebuilder will pass all flags to the plugin, making it the external plugin's responsibility to handle any invalid flags. 

> If a plugin does implement the `flags` subcommand, it **MUST** support `--init`, `--api`, `--webhook`, and `--edit` flags, which correspond to respective Kubebuilder subcommands.

Reviewer: Nit: Now that I am thinking about it a bit more I don't think these flags are _required_ to be implemented as the external plugin would return an error about an unknown flag that Kubebuilder would ignore. That being said, we may want to consider the pros and cons of enforcing or not enforcing this behavior. 

@Eileen-Yu I'll leave it up to you if you want to change the wording to reflect that these flags aren't _required_ . IMO though this nit shouldn't block this PR from merging.

Author: I can understand that flag for `--init` … is more like a ‘bonus’ instead of ‘must’. For me, actually I’m still curious about the detailed scenario to support that so as to reach “fast fail”.

So I guess maybe we can drop that paragraph includes the 'must' literal.
I was imagining if it is fair enough to stay focus on the general usage(`init`, `create` and `edit`) of the external plugin.

# Extending Kubebuilder with external plugins

## Overview

Kubebuilder's functionality can be extended through the use of external plugins.

An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language. 

> The Kubebuilder CLI loads the external plugin in the specified path and interact the plugin through the executable's `stdin` & `stdout`.

Reviewer: Nit: Some suggestions for wording improvement:

```suggestion
The Kubebuilder CLI loads the external plugin in the specified path and interacts with it through `stdin` & `stdout`.
```

## When is it useful?

- If you want to create helpers or addons on top of the scaffolds done by Kubebuilder's default scaffolding.

- If you design customized layouts and want to take advantage of functions from Kubebuilder library.

- If you are looking for implementing plugins in a language other than `Go`.

## How to write it?

The inter-process communication between your external plugin and Kubebuilder is through the standard I/O.

> You can write your external plugin in any language as long as it implements the `PluginRequest` and `PluginResponse` interfaces. 

Reviewer: Nit: "interfaces" suggests a looser construct in programming ("interfaces" can be anything behind the scenes as long as they implement certain functionality). I would recommend using "types" as it is more strict and aligns with what the external plugin communication process in Kubebuilder expects. The response must be exactly the `PluginResponse` type and shouldn't have any additional fields

Author: Thank you for the nice explanation! 👍
I tried to reorganized this paragraph in the latest commit.

	// Error is a boolean type that indicates whether there were any errors due to plugin failures.
	Error bool `json:"error,omitempty"`

	// ErrorMsgs contains the specific error messages of the plugin failures.
	ErrorMsgs []string `json:"errorMsgs,omitempty"`

	// Flags contains the plugin specific flags that the plugin returns to Kubebuilder when it receives
	// a request for a list of supported flags from Kubebuilder
	Flags []Flag `json:"flags,omitempty"`
}
```

<aside class="note">
<h1>Caution</h1>

When writing your own external plugin, you **should not** directly echo or print anything to the stdout. 

This is because Kubebuilder and your plugin are communicating with each other via `stdin` and `stdout` using structured `JSON` data.
Any additional information sent to stdout (such as debug messages or logs) that's not part of the expected PluginResponse JSON structure may cause parsing errors when Kubebuilder tries to read and decode the response from your plugin.

If you need to include logs or debug messages while developing your plugin, consider writing these messages to a log file instead.

</aside>

## How to use it?

### Prerequisites
- kubebuilder CLI > 3.11.0

> - Your own external plugin executable written in a language of your choice 

Reviewer: I'm not sure this should be a prerequisite - I'm interpreting this as I _have_ to write my own external plugin rather than being able to use any external plugin

Author: Oh I see!
So by this line I tried to say the external plugin needs to be an executable..
I updated this part with some explanations.

- Path of the external plugin configured via the ENV VAR `${EXTERNAL_PLUGINS_PATH}` or it can be placed in a path following a `group-like name and version` scheme:
```sh
# for Linux
$HOME/.config/kubebuilder/plugins/${name}/${version}

# for OSX
~/Library/Application Support/kubebuilder/plugins/${name}/${version}
```

### Subcommands:

The external plugin supports the same subcommands as kubebuilder already provides:
- `init`: project initialization.
- `create api`: scaffold Kubernetes API definitions.
- `create webhook`: scaffold Kubernetes webhooks.
- `edit`: update the project configuration.

Also, there are **Optional** subcommands for a better user experience:
- `metadata`: add customized plugin description and examples when a `--help` flag is specified. 
- `flags`: specify valid flags for Kubebuilder to pass to the external plugin.

<aside class="note">
<h1>More about `flags` subcommand</h1>

> The flags subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 

Reviewer: Nit:
```suggestion
The `flags` subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
```

> If a plugin does not implement the flags subcommand, Kubebuilder will pass all flags to the plugin, which then it is the external plugin's responsibility to handle those invalid flags. 

Reviewer: nit:
```suggestion
If a plugin does not implement the `flags` subcommand, Kubebuilder will pass all flags to the plugin, making it the external plugin's responsibility to handle any invalid flags. 
```

> If a plugin does implement the flags subcommand, it **MUST** support `init`, `create api`, `create webhook`, and `edit` flags, corresponding to respective Kubebuilder subcommands.

Reviewer: nit:
```suggestion
If a plugin does implement the `flags` subcommand, it **MUST** support `--init`, `--api`, `--webhook`, and `--edit` flags, which correspond to respective Kubebuilder subcommands.
```
